README.md
e4x.js
======

**e4x.js** implements the optional E4X features described in ECMA-357 2nd Edition Annex A.


jQuery E4X DOM Plugin
=====================

The [jQuery E4X DOM Plugin][1] ([jquery.e4x-dom.js][2]) adds support for XML nodes as DOM
nodes in jQuery. It also adds support for converting jQuery objects into XMLLists.
If an XML element does not have a namespace, it is assigned the XHTML 1.x namespace.
To change the default assigned namespace, change `jQuery.defaultXMLNamespace` accordingly.



![Tracking image](//in.getclicky.com/212712ns.gif =1x1)


 [1]: http://plugins.jquery.com/project/e4x-dom
 [2]: http://purl.eligrey.com/github/e4x.js/blob/master/jquery-plugin/jquery.e4x-dom.js


changelog.md
2010-06-30
==========

* Now using `document.implementation.createDocument` instead of `DOMParser` to create
  DOM documents.
* Tidied up the XML prototype method extension code.


2010-05-27
==========

* Minor speed and minification optimizations.


2010-04-02
==========

* Fixed `methods` variable global leak.
* Removed checks for `DOMParser` and `XMLSerializer`.


2010-03-13
==========

* No longer using version numbers as e4x.js only implements a static feature set.


0.2.2
=====

* Now partially using DOM methods to create a DOM document instead of using string
  serialization for the `xpath()` method.


0.2.1
=====

* Optimizations.


jquery-plugin/jquery.e4x-dom.js
/*
 * jQuery E4X DOM Plugin
 * Version 0.1.5
 * 2010-06-30
 * 
 * The jQuery E4X DOM plugin adds support of interchanging E4X XML and DOM nodes
 * with jQuery through a modified initialization method that converts XML to DOM nodes
 * and a jQuery.fn.xml method that converts DOM nodes to XML. 
 * 
 * By Eli Grey, http://eligrey.com
 * License: The X11/MIT license
 */

/*jslint evil: true, strict: true, undef: true, nomen: true, eqeqeq: true, bitwise: true,
  regexp: true, newcap: true, immed: true, maxerr: 1000, maxlen: 90 */

/*global jQuery, XML, XMLList, XMLSerializer, document */

"use strict";

if (typeof XML !== "undefined") {

(function (XML, $, xmlSerializer) {
	var Init = $.fn.init,
	hostDoc  = document,
	NULL     = null,
	xmlDoc   = hostDoc.implementation.createDocument(NULL, NULL, NULL),
	piName   = /^[\w\-]+\s*/,
	defaultXMLNSProp = "defaultXMLNamespace",
	domNodeList = XML.prototype.function::domNodeList;
	
	if (!domNodeList) {
		var domNode = function (xml) {
			var node;
		
			switch (xml.nodeKind()) {
				case "element":
					var attributes = xml.attributes(),
						children   = xml.children(),
						childLen   = children.length(),
						attLen     = attributes.length(),
						i, attribute;
				
					node = xmlDoc.createElementNS(
						xml.namespace().uri || $[defaultXMLNSProp],
						xml.localName()
					);
				
					if (attLen !== 0) {
						for (i = 0; i < attLen; i++) {
							attribute = attributes[i];
							node.setAttributeNS(
								attribute.namespace().uri,
								attribute.localName(),
								attribute.toString()
							);
						}
					}
					if (childLen !== 0) {
						for (i = 0; i < childLen; i++) {
							node.appendChild(domNode(children[i]));
						}
					}
				
					return node;
			
				case "text":
					return xmlDoc.createTextNode(xml.toString());
			
				case "comment":
					return xmlDoc.createComment(xml.toString().slice(4, -3));
			
				case "processing-instruction":
					return xmlDoc.createProcessingInstruction(
						xml.localName(),
						xml.toString().slice(2, -2).replace(piName, "")
					);
			
				case "attribute":
					(node = xmlDoc.createAttributeNS(
						xml.namespace().uri,
						xml.localName()
					)).nodeValue = xml.toString();
					return node;
			}
		};
		domNodeList = function () {
			var fragment = hostDoc.createDocumentFragment();
		
			for (var i = 0, len = this.length(); i < len; i++) {
				fragment.appendChild(hostDoc.adoptNode(domNode(this[i])));
			}
		
			return hostDoc.adoptNode(fragment).childNodes;
		};
	}
	
	// If XML elements don't have a namespace, they will be set to the XHTML 1.x namespace
	$[defaultXMLNSProp] = "http://www.w3.org/1999/xhtml";
	
	$.fn.init = function (selector, context) {
		return new Init(typeof selector === "xml" ? domNodeList.call(selector) : selector,
		                typeof context  === "xml" ? domNodeList.call(context)  : context);
	};
	
	$.fn.xml = function () {
		var xml = new XMLList,
		    i   = 0,
		    len = this.length;
		
		for (; i < len; i++) {
			// Unfortunately, there's no efficient native XML.fromDomNode(node) method
			xml += new XML(xmlSerializer.serializeToString(this[i]));
		}
		
		return xml;
	};
	
	XML.ignoreWhitespace = false;
}(XML, jQuery, new XMLSerializer));

}


e4x.js
/*
 * e4x.js
 * 
 * A JavaScript library that implements the optional E4X features described in
 * ECMA-357 2nd Edition Annex A if they are not already implemented.
 *
 * 2010-06-30
 * 
 * By Eli Grey, http://eligrey.com
 * License: The X11/MIT license (see COPYING.md)
 */

/*global document, XML, XMLList, XMLSerializer, XPathResult */

/*jslint undef: true, nomen: true, eqeqeq: true, bitwise: true, regexp: true,
  newcap: true, immed: true, maxerr: 1000, maxlen: 90 */

"use strict";

(function (XML) {
	var hostDoc   = document,
	xmlSerializer = new XMLSerializer,
	piName        = /^[\w\-]+\s*/,
	NULL          = null,
	createDoc     = function (docElem) {
		var domDoc = hostDoc.implementation.createDocument(NULL, NULL, NULL);
		if (docElem) {
			domDoc.appendChild(docElem);
		}
		return domDoc;
	},
	xmlDoc        = createDoc(),
	
	xmlToDomNode = function (xml) {
		var node;
		
		switch (xml.nodeKind()) {
			case "element":
				var attributes = xml.attributes(),
				    children   = xml.children(),
				    childLen   = children.length(),
				    attLen     = attributes.length(),
				    i, attribute;
				
				node = xmlDoc.createElementNS(
					xml.namespace().uri,
					xml.localName()
				);
				
				if (attLen !== 0) {
					for (i = 0; i < attLen; i++) {
						attribute = attributes[i];
						node.setAttributeNS(
							attribute.namespace().uri,
							attribute.localName(),
							attribute.toString()
						);
					}
				}
				if (childLen !== 0) {
					for (i = 0; i < childLen; i++) {
						node.appendChild(xmlToDomNode(children[i]));
					}
				}
				
				return node;
			
			case "text":
				return xmlDoc.createTextNode(xml.toString());
			
			case "comment":
				return xmlDoc.createComment(xml.toString().slice(4, -3));
			
			case "processing-instruction":
				return xmlDoc.createProcessingInstruction(
					xml.localName(),
					xml.toString().slice(2, -2).replace(piName, "")
				);
			
			case "attribute":
				(node = xmlDoc.createAttributeNS(
					xml.namespace().uri,
					xml.localName()
				)).nodeValue = xml.toString();
				return node;
		}
	},
	xmlMethods = {
		domNode: function () {
			if (this.length() === 1) {
				return hostDoc.adoptNode(xmlToDomNode(this));
			}
		},
		domNodeList: function () {
			var fragment = hostDoc.createDocumentFragment();
		
			for (var i = 0, len = this.length(); i < len; i++) {
				fragment.appendChild(this[i].domNode());
			}
		
			return hostDoc.adoptNode(fragment).childNodes;
		},
		xpath: function (xpathExp) {
			var res = new XMLList,
			    len = this.length();
	
			if (len !== 1) {
				for (var i = 0, len = len; i < len; i++) {
					res += this[i].xpath(xpathExp);
				}
				return res;
			}
	
			var domDoc = createDoc(this.domNode()),
			xpr = domDoc.evaluate(
				xpathExp,
				domDoc.documentElement,
				domDoc.createNSResolver(domDoc.documentElement),
				XPathResult.ORDERED_NODE_ITERATOR_TYPE,
				NULL
			),
			node;
	
			while (node = xpr.iterateNext()) {
				// Unfortunately, there's no efficient native XML.fromDomNode(node) method
				res += new XML(xmlSerializer.serializeToString(node));
			}
	
			return res;
		}
	},
	method;
	
	for (method in xmlMethods) {
		if (xmlMethods.hasOwnProperty(method) && !XML.prototype.function::[method]) {
			XML.prototype.function::[method] = xmlMethods[method];
		}
	}
}(XML));
